fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ExternalTablePredicateExtractor.java

```diff
@@ -0,0 +1,139 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+package com.starrocks.sql.optimizer.rewrite;
+
+import com.starrocks.sql.optimizer.Utils;
+import com.starrocks.sql.optimizer.operator.OperatorType;
+import com.starrocks.sql.optimizer.operator.scalar.BetweenPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.BinaryPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.CastOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.CompoundPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ConstantOperator;
+import com.starrocks.sql.optimizer.operator.scalar.InPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.IsNullPredicateOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperatorVisitor;
+
+import java.util.LinkedList;
+import java.util.List;
+
+// Extract predicates that can be pushed down to external table
+// and predicates that must be reserved
+// from the entire predicate
+// To be safe, we only allow push down simple  predicates
+public class ExternalTablePredicateExtractor {
+    private List<ScalarOperator> pushedPredicates = new LinkedList<>();
+    private List<ScalarOperator> reservedPredicates = new LinkedList<>();
+
+    public ScalarOperator getPushPredicate() {
+        return Utils.compoundAnd(pushedPredicates);
+    }
+
+    public ScalarOperator getReservePredicate() {
+        return Utils.compoundAnd(reservedPredicates);
+    }
+
+    public void extract(ScalarOperator op) {
+        pushedPredicates.clear();
+        reservedPredicates.clear();
+
+        if (op.getOpType().equals(OperatorType.COMPOUND)) {
```

Comment:
CanFullyPushDownVisitor and here both deal with CompoundPredicate， it's looks like have some problems, if predicate like ((canPush) and (can not Push)) and canPush, it will lose one canPush predicate?

Reply:
> CanFullyPushDownVisitor and here both deal with CompoundPredicate， it's looks like have some problems, if predicate like ((canPush) and (can not Push)) and canPush, it will lose one canPush predicate?

@Youngwb  yes…but before using this rule，all `AND` predicates should have been flattened into a CompoundAndPredicate? This is an important assumption.

Reply:
No, it have not been  flattened

Reply:
> No, it have not been flattened

@Youngwb fixed